### PR TITLE
[Merged by Bors] - feat(Order/InitSeg): notation for initial and principal segments of `<`

### PR DESCRIPTION
**Pull request by vihdzp**
We introduce the notations `α ≤i β` and `α <i β` for initial/principal segment embeddings between the `<` relations of two types. We also prove a few lemmas.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Order/InitialSeg.lean
Comment (YaelDillies) [inline Mathlib/Order/InitialSeg.lean:62]: Why make it an abbrev?
```suggestion
/-- An `InitialSeg` between the `<` relations of two types. -/
notation:25 alpha:24 " ≤i " beta:25 => @InitialSegLT α β (· < ·) (· < ·)
```
or something like that

File: Mathlib/Order/InitialSeg.lean
Comment (YaelDillies) [inline Mathlib/Order/InitialSeg.lean:89]:
```suggestion
@[simp, norm_cast]
```

File: Mathlib/Order/InitialSeg.lean
Comment (YaelDillies) [inline Mathlib/Order/InitialSeg.lean:235]: Can you check gcongr proves the <- direction? If it doesn't, you might need to add an `OrderHomClass` instance

File: Mathlib/Order/InitialSeg.lean
Comment (YaelDillies) [inline Mathlib/Order/InitialSeg.lean:31]: Can you discuss in the module doc why this pair of notations makes sense for initial and principal segments?

File: Mathlib/Order/InitialSeg.lean
Comment (vihdzp) [reply Mathlib/Order/InitialSeg.lean:62]: Nice! I'm not really familiar with the `notation` syntax, but this is much cleaner.

File: Mathlib/Order/InitialSeg.lean
Comment (vihdzp) [reply Mathlib/Order/InitialSeg.lean:86]: "badly shaped lemma, rhs can't start with coe"

File: Mathlib/Order/InitialSeg.lean
Comment (vihdzp) [reply Mathlib/Order/InitialSeg.lean:235]: I added the `OrderHomClass` instance, but this still makes no progress:
```
example [PartialOrder α] (f : α ≤i β) (h : a ≤ a') : f a ≤ f a' := by gcongr
```

File: Mathlib/Order/InitialSeg.lean
Comment (vihdzp) [reply Mathlib/Order/InitialSeg.lean:86]: "badly shaped lemma, rhs can't start with coe"

File: Mathlib/Order/InitialSeg.lean
Comment (vihdzp) [reply Mathlib/Order/InitialSeg.lean:235]: I added the `OrderHomClass` instance, but this still makes no progress:
```lean 4
example [PartialOrder α] (f : α ≤i β) (h : a ≤ a') : f a ≤ f a' := by gcongr
```

File: Mathlib/Order/InitialSeg.lean
Comment (vihdzp) [reply Mathlib/Order/InitialSeg.lean:31]: I just kinda came up with it. My reasoning is that `≼` and `≺` refer to "general" relations, while `≤` and `<` are linked to the order typeclass hierarchy.

Comment (YaelDillies) [reply Mathlib/Order/InitialSeg.lean:31]: Sure, but why is "initial segment" the non-strict version of "principal segment"?

Comment (vihdzp) [reply Mathlib/Order/InitialSeg.lean:31]: Because of [`InitialSeg.ltOrEq`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/InitialSeg.html#InitialSeg.ltOrEq). Do I link that lemma in this notation section?

File: Mathlib/Order/InitialSeg.lean
Comment (vihdzp) [reply Mathlib/Order/InitialSeg.lean:235]: Did you happen to add it in #17224? Because that PR is open :P

File: Mathlib/Order/InitialSeg.lean
Comment (vihdzp) [reply Mathlib/Order/InitialSeg.lean:31]: I've added a small comment referring to this theorem.

File: Mathlib/Order/InitialSeg.lean
Comment (YaelDillies) [reply Mathlib/Order/InitialSeg.lean:238]: Ah damn!

**Comment by YaelDillies** [comment]
Sorry, I see I commented on the wrong PR. See the dependent PR.

**Review by YaelDillies — APPROVED** [review]
maintainer merge

File: Mathlib/Order/InitialSeg.lean
Comment (b-mehta) [inline Mathlib/Order/InitialSeg.lean:53]: This note still applies, I'm pretty sure

File: Mathlib/Order/InitialSeg.lean
Comment (b-mehta) [inline Mathlib/Order/InitialSeg.lean:222]: as before

File: Mathlib/Order/InitialSeg.lean
Comment (b-mehta) [inline Mathlib/Order/InitialSeg.lean:95]: In the general (not-partial-order) case, we also have RelHomClass here, right?

File: Mathlib/Order/InitialSeg.lean
Comment (b-mehta) [inline Mathlib/Order/InitialSeg.lean:234]: Seems like this and the below should all follow nicely if we had RelEmbeddingClass... Let's add a TODO to this effect

File: Mathlib/Order/InitialSeg.lean
Comment (b-mehta) [inline Mathlib/Order/InitialSeg.lean:292]: Don't you want all the le_iff_le; lt_iff_lt lemmas here too?

File: Mathlib/Order/InitialSeg.lean
Comment (vihdzp) [reply Mathlib/Order/InitialSeg.lean:53]: Is it still useful though? As I understand, the notation should be scoped even without this.

Comment (YaelDillies) [reply Mathlib/Order/InitialSeg.lean:53]: What makes you think the notation is currently scoped?

Comment (vihdzp) [reply Mathlib/Order/InitialSeg.lean:53]: I recall having to open the namespace to have this notation work. If it's not scoped, should it be?

File: Mathlib/Order/InitialSeg.lean
Comment (vihdzp) [reply Mathlib/Order/InitialSeg.lean:292]: Nice catch! Yeah, I forgot about these. Fortunately it's just a bunch of one liners, so I'll bunch all of these theorems together at the end of the file.

File: Mathlib/Order/InitialSeg.lean
Comment (YaelDillies) [reply Mathlib/Order/InitialSeg.lean:53]: It doesn't seem scoped to me at all. I don't know whether it should be, but certainly it's out of scope (🥁) to scope it or to remove the porting in this PR

File: Mathlib/Order/InitialSeg.lean
Comment (vihdzp) [reply Mathlib/Order/InitialSeg.lean:86]: It doesn't seem to work for whatever reason.

File: Mathlib/Order/InitialSeg.lean
Comment (vihdzp) [reply Mathlib/Order/InitialSeg.lean:95]: Yep, I've opened #17598 where I add this alongside some missing instances on `PrincipalSeg`.

File: Mathlib/Order/InitialSeg.lean
Comment (b-mehta) [inline Mathlib/Order/InitialSeg.lean:59]:
```suggestion
-- Porting note: Deleted `scoped[InitialSeg]`
```
very minor but I think this might make git assign the blame for this line correctly

**Comment by b-mehta** [comment]
lgtm, ping me once the merge conflict is fixed and I'll hit merge

**Comment by vihdzp** [comment]
I think it should be fixed now, but let's wait for CI just in case.

**Review by b-mehta — COMMENTED** [review]
Thanks!

bors merge

**Pull request closed without merging**
